go/database/mpt/proof/verification_test.go

```diff
@@ -270,61 +332,239 @@ func TestVerification_FailingLiveTrie(t *testing.T) {
+		t.Fatalf("failed to verify archive trie: %v", err)
+	}
+
+	// exercise the mock based on the number of executions
+	loops := counter
+	for i := 0; i < loops; i++ {
+		counter = 0   // reset the counter
+		threshold = i // update the threshold every loop
+		if err := verifyTrie(context.Background(), errorInjectingTrieMock, observer); !errors.Is(err, ErrInvalidProof) {
+			t.Errorf("expected error %v, got %v", ErrInvalidProof, err)
+		}
+	}
+}
+
+func TestVerification_VerifyProof_Incomplete_Or_Empty(t *testing.T) {
+	tests := []struct {
+		name           string
+		returnComplete bool
+	}{
+		{"incomplete", false},
+		{"empty", true},
+	}
+
 	address := common.Address{1}
 	data := mpt.AccountInfo{Nonce: common.Nonce{1}, Balance: amount.New(1), CodeHash: common.Hash{1}}
 	storage := map[common.Key]common.Value{{1}: {1}}
 
-	ctrl := gomock.NewController(t)
-	parent := witness.NewMockProof(ctrl)
-	parent.EXPECT().GetState(gomock.Any(), gomock.Any(), gomock.Any()).Return(common.Value{1}, true, nil).AnyTimes()
-	parent.EXPECT().GetBalance(gomock.Any(), gomock.Any()).Return(amount.New(1), true, nil).AnyTimes()
-	parent.EXPECT().GetNonce(gomock.Any(), gomock.Any()).Return(common.Nonce{1}, true, nil).AnyTimes()
-	parent.EXPECT().GetCodeHash(gomock.Any(), gomock.Any()).Return(common.Hash{1}, true, nil).AnyTimes()
-
-	var x int
-	proof := errorInjectingProof{Proof: parent, returnComplete: true, returnError: nil, threshold: 1000, transientCount: &x}
-	if err := verifyAccount(common.Hash{}, &proof, address, data); err != nil {
-		t.Errorf("expected no error, got %v", err)
-	}
-
-	t.Run("incomplete proofs", func(t *testing.T) {
-		for i := 0; i < proof.count; i++ {
-			proof := errorInjectingProof{Proof: parent, returnComplete: false, returnError: nil, threshold: i, transientCount: &x}
-			if err := verifyAccount(common.Hash{}, &proof, address, data); err == nil {
-				t.Errorf("expected error, got nil")
-			}
-			if err := verifyStorage(common.Hash{}, &proof, address, []common.Key{{1}}, storage); err == nil {
-				t.Errorf("expected error, got nil")
-			}
-		}
-	})
-	t.Run("empty values in proofs", func(t *testing.T) {
-		for i := 0; i < proof.count; i++ {
-			proof := errorInjectingProof{Proof: parent, returnComplete: true, returnError: nil, threshold: i, transientCount: &x}
-			if err := verifyAccount(common.Hash{}, &proof, address, data); err == nil {
-				t.Errorf("expected error, got nil")
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			var count int
+			threshold := 1000
+			errorInjectingProofMock := witness.NewMockProof(ctrl)
+			errorInjectingProofMock.EXPECT().GetBalance(gomock.Any(), gomock.Any()).DoAndReturn(func(root common.Hash, addr common.Address) (amount.Amount, bool, error) {
+				if count >= threshold {
+					return amount.Amount{}, test.returnComplete, nil
+				}
+				count++
+				return amount.New(1), true, nil
+			}).AnyTimes()
```

Comment:
I would have thought something like
```
gomock.InOrder(
   errorInjectingProofMock.EXPECT().GetBalance(...).Times(theshold).Return(amount.New(1), true, nil),
   errorInjectingProofMock.EXPECT().GetBalance(...).AnyTimes().Return(amount.Amount{}, test.returnComplete, nil),
)
```
would have done the trick here. But this depends on whether for this test here it is essential that all expectations share a common counter.

Reply:
do I understand this idea correctly that I would need to manually check `threshold` for each function, and set it accordingly for each method?